// appium-xcuitest-driver/build/lib/commands/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatteryState = exports.AppState = void 0;
var AppState;
(function (AppState) {
    /**
     * The application’s current state is not known.
     */
    AppState[AppState["XCUIApplicationStateUnknown"] = 0] = "XCUIApplicationStateUnknown";
    /**
     * The application is not running
     */
    AppState[AppState["XCUIApplicationStateNotRunning"] = 1] = "XCUIApplicationStateNotRunning";
    /**
     * The application is running in the background, but is suspended.
     */
    AppState[AppState["XCUIApplicationStateRunningBackgroundSuspended"] = 2] = "XCUIApplicationStateRunningBackgroundSuspended";
    /**
     * The application is running in the background.
     */
    AppState[AppState["XCUIApplicationStateRunningBackground"] = 3] = "XCUIApplicationStateRunningBackground";
    /**
     * The application is running in the foreground.
     */
    AppState[AppState["XCUIApplicationStateRunningForeground"] = 4] = "XCUIApplicationStateRunningForeground";
})(AppState = exports.AppState || (exports.AppState = {}));
/**
 * Battery state
 * @see {@linkcode BatteryInfo}
 */
var BatteryState;
(function (BatteryState) {
    BatteryState[BatteryState["UIDeviceBatteryStateUnknown"] = 0] = "UIDeviceBatteryStateUnknown";
    BatteryState[BatteryState["UIDeviceBatteryStateUnplugged"] = 1] = "UIDeviceBatteryStateUnplugged";
    BatteryState[BatteryState["UIDeviceBatteryStateCharging"] = 2] = "UIDeviceBatteryStateCharging";
    BatteryState[BatteryState["UIDeviceBatteryStateFull"] = 3] = "UIDeviceBatteryStateFull";
})(BatteryState = exports.BatteryState || (exports.BatteryState = {}));
