// appium-webdriveragent/build/lib/constants.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WDA_UPGRADE_TIMESTAMP_PATH = exports.WDA_SCHEME = exports.WDA_RUNNER_BUNDLE_ID_FOR_XCTEST = exports.WDA_RUNNER_BUNDLE_ID = exports.WDA_RUNNER_APP = exports.WDA_BASE_URL = exports.SDK_SIMULATOR = exports.SDK_DEVICE = exports.PROJECT_FILE = exports.PLATFORM_NAME_TVOS = exports.PLATFORM_NAME_IOS = void 0;
require("source-map-support/register");
var _path = _interopRequireDefault(require("path"));
const WDA_RUNNER_BUNDLE_ID = 'com.facebook.WebDriverAgentRunner';
exports.WDA_RUNNER_BUNDLE_ID = WDA_RUNNER_BUNDLE_ID;
const WDA_RUNNER_BUNDLE_ID_FOR_XCTEST = `${WDA_RUNNER_BUNDLE_ID}.xctrunner`;
exports.WDA_RUNNER_BUNDLE_ID_FOR_XCTEST = WDA_RUNNER_BUNDLE_ID_FOR_XCTEST;
const WDA_RUNNER_APP = 'WebDriverAgentRunner-Runner.app';
exports.WDA_RUNNER_APP = WDA_RUNNER_APP;
const WDA_SCHEME = 'WebDriverAgentRunner';
exports.WDA_SCHEME = WDA_SCHEME;
const PROJECT_FILE = 'project.pbxproj';
exports.PROJECT_FILE = PROJECT_FILE;
const WDA_BASE_URL = 'http://127.0.0.1';
exports.WDA_BASE_URL = WDA_BASE_URL;
const PLATFORM_NAME_TVOS = 'tvOS';
exports.PLATFORM_NAME_TVOS = PLATFORM_NAME_TVOS;
const PLATFORM_NAME_IOS = 'iOS';
exports.PLATFORM_NAME_IOS = PLATFORM_NAME_IOS;
const SDK_SIMULATOR = 'iphonesimulator';
exports.SDK_SIMULATOR = SDK_SIMULATOR;
const SDK_DEVICE = 'iphoneos';
exports.SDK_DEVICE = SDK_DEVICE;
const WDA_UPGRADE_TIMESTAMP_PATH = _path.default.join('.appium', 'webdriveragent', 'upgrade.time');
exports.WDA_UPGRADE_TIMESTAMP_PATH = WDA_UPGRADE_TIMESTAMP_PATH;
