// @appium/test-support/build/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySandbox = exports.withSandbox = exports.verifyMocks = exports.withMocks = exports.fakeTime = exports.stubLog = exports.stubEnv = void 0;
// this just needs to be imported, for the functionality to be injected
require("./unhandled-rejection");
var env_utils_1 = require("./env-utils");
Object.defineProperty(exports, "stubEnv", { enumerable: true, get: function () { return env_utils_1.stubEnv; } });
var log_utils_1 = require("./log-utils");
Object.defineProperty(exports, "stubLog", { enumerable: true, get: function () { return log_utils_1.stubLog; } });
var time_utils_1 = require("./time-utils");
Object.defineProperty(exports, "fakeTime", { enumerable: true, get: function () { return time_utils_1.fakeTime; } });
var mock_utils_1 = require("./mock-utils");
Object.defineProperty(exports, "withMocks", { enumerable: true, get: function () { return mock_utils_1.withMocks; } });
Object.defineProperty(exports, "verifyMocks", { enumerable: true, get: function () { return mock_utils_1.verifyMocks; } });
var sandbox_utils_1 = require("./sandbox-utils");
Object.defineProperty(exports, "withSandbox", { enumerable: true, get: function () { return sandbox_utils_1.withSandbox; } });
Object.defineProperty(exports, "verifySandbox", { enumerable: true, get: function () { return sandbox_utils_1.verifySandbox; } });
